src/sanitation.js

```diff
@@ -241,21 +250,56 @@ export function isValidAttr(tagName, attrName, attrValue, opt_purify = false) {
   }
 
   // Remove blacklisted attributes from specific tags e.g. input[formaction].
-  const attrNameBlacklist = BLACKLISTED_TAG_SPECIFIC_ATTRS[tagName];
-  if (attrNameBlacklist && attrNameBlacklist.indexOf(attrName) != -1) {
+  let attrBlacklist = BLACKLISTED_TAG_SPECIFIC_ATTRS[tagName] || [];
+  if (isAmp4Email(doc)) {
+    const amp4EmailAttrBlacklist =
+        {'form': ['name'], 'amp-anim': ['controls']}[tagName] || [];
+    attrBlacklist = attrBlacklist.concat(amp4EmailAttrBlacklist);
+  }
+  if (attrBlacklist.indexOf(attrName) != -1) {
     return false;
   }
 
   // Remove blacklisted values for specific attributes for specific tags
   // e.g. input[type=image].
-  const attrBlacklist = BLACKLISTED_TAG_SPECIFIC_ATTR_VALUES[tagName];
-  if (attrBlacklist) {
-    const blacklistedValuesRegex = attrBlacklist[attrName];
-    if (blacklistedValuesRegex &&
-        attrValue.search(blacklistedValuesRegex) != -1) {
-      return false;
+  const attrValueBlacklist =
+      BLACKLISTED_TAG_SPECIFIC_ATTR_VALUES[tagName] || dict();
+  let blacklistedValues = attrValueBlacklist[attrName] || [];
+  if (isAmp4Email(doc)) {
+    const amp4EmailAttrValueBlacklist =
+        AMP4EMAIL_BLACKLISTED_TAG_SPECIFIC_ATTR_VALUES[tagName];
+    if (amp4EmailAttrValueBlacklist) {
+      const blacklistAmp4EmailValues =
+          amp4EmailAttrValueBlacklist[attrName] || [];
+      blacklistedValues = blacklistedValues.concat(blacklistAmp4EmailValues);
```

Comment:
Why change to string arrays and `concat()` instead of just replacing the original regex? 

Replacing the regex would (1) be simpler code and (2) be closer to the validator rule and (3) not require `toLowerCase()` calls: https://github.com/ampproject/amphtml/blob/ffe56554e7a6711d789a768ab1cf8689301dd216/validator/validator-main.protoascii#L4420-L4432

Reply:
Ok but one can also make the argument that with this change it is in par with how the other blacklists and whitelists are done in this file, which are all arrays. The reason for the array is that I am extending the blacklisted tag specific attributes with the additional values if AMP4Email. I can switch back to a regex but that would mean either creating a new regex with the duplicated values from BLACKLISTED_TAG_SPECIFIC_ATTR_VALUES, which is what I had originally.

e.g.

    /** @const {!Object<string, !Object<string, !RegExp>>} */
    const AMP4EMAIL_BLACKLISTED_TAG_SPECIFIC_ATTR_VALUES = dict({
      'input': {
      'type': /(?:image|button|file|password)/i,
      },
    });

and that would mean, being mindful that if BLACKLISTED_TAG_SPECIFIC_ATTR_VALUES is modified, we need to consider making the change to the AMP4EMAIL_BLACKLISTED_TAG_SPECIFIC_ATTR_VALUES  regex as well.

Or creating a regex expression dynamically with the values from two lists, which would mean changing the values from both lists so that they are strings, so that we can do the something like the  following:
    
    const amp4EmailBlacklistRegex = new RegExp(BLACKLISTED_TAG_SPECIFIC_ATTR_VALUES[tagName] + 
         AMP4EMAIL_BLACKLISTED_TAG_SPECIFIC_ATTR_VALUES[tagName]);

Reply:
I think duplication is fine since it mirrors how the validator actually works (it doesn't merge rulesets) and results in simpler code.